### ISSUE TITLE
reduce t2t nmt model size

### ISSUE THREAD

**Issue by Mr-wang2016** [tensorflow/tensor2tensor#1482]
Hi:

   I trained a nmt model use t2t  with vocab size 32000, the nmt model size is about 1.2G, I think it is a little big, so how can i reduce the size of my model with little loss about blue? Any Suggestions would be appreciated, thanks.

**Comment by martinpopel** [comment]
I guess this is `transformer_big`. You can try `transformer_base`, which may result in lower BLEU. You can also try Afafactor instead of Adam, which should make the model about 60% smaller.

**Comment by Mr-wang2016** [comment]
Sorry I didn't describe it clearly. the hparams_set i use is 'transformer_base_single_gpu'. when i use transformer_big, the nmt model size will be 1.6G and i did not have a try with transformer_base. Will Afafactor result in lower BLEU then Adam?thanks for responding.

**Comment by martinpopel** [comment]
The `single_gpu` suffix should not affect the model size. In my experience Adafactor gives better results than Adam because I can afford higher batch size, but I need to use Adafactor with `learning_rate_schedule=rsqrt_decay` or `learning_rate_schedule=linear_warmup*rsqrt_decay` (depending on T2T version).

**Comment by Mr-wang2016** [comment]
thanks @martinpopel , i will have a try.

**Comment by ndvbd** [comment]
> I guess this is `transformer_big`. You can try `transformer_base`, which may result in lower BLEU. You can also try Afafactor instead of Adam, which should make the model about 60% smaller.

@martinpopel Just to clarify, you mean that the model footprint during training (both memory and checkpoint) will be 60%, but not the final exported model for inference (which will be identical).

**Comment by martinpopel** [comment]
> you mean that the model footprint during training (both memory and checkpoint) will be 60%

Yes.

> not the final exported model for inference (which will be identical)

I don't remember if the default export procedure removes the first and second moments (for Adam/Adafactor). If yes, then the final models will have the same size. If not, you could probably remove these moments manually (their names starts with `training/`). The models won't have identical weights because Adafactor is a different optimizer than Adam, but in my experience the BLEU results are very similar if training with the same batch size (but Adafactor allows you to train with a higher batch size).

Yet another way how to train smaller models without losing much quality (BLEU) is the standard model distillation: translate big monolingual data as well as the source side of your parallel data with the big teacher model. Then use this newly created parallel corpus for training a small student model - you can e.g. decrease the number of layers in the decoder to just two (or replace the Transformer in decoder with some kind of GRU). For details see e.g. [Microsoft's excellent work](https://www.aclweb.org/anthology/D19-5632.pdf) with many other practical tricks.

**Comment by ndvbd** [comment]
@martinpopel Excellent. So if the Adafactor is better than Adam, we don't need Adam anymore to be the default in t2t, do we?
Regarding the tricks you mentioned, is there an easy way to quantize the model for inference, for example float -> 8 bit quantization?

**Comment by martinpopel** [comment]
> So if the Adafactor is better than Adam, we don't need Adam anymore to be the default in t2t, do we?

Adam is good for comparison with prior work and other frameworks, which don't have Adafactor, btw. there is [no TF2 Adafactor implementation](https://github.com/tensorflow/addons/issues/522) yet. There maybe tasks where memory is not a problem (either you have large GPUs or larger batch size is not beneficial for a given tasks) or tasks where Adafactor does not work - I don't know; I have tried only MT. Also, there is MultistepAdamOptimizer in T2T (now in [two](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/utils/multistep_optimizer.py) [implementations](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/utils/multistep_with_adamoptimizer.py) for different T2T+TF versions), which can be used to increase the effective batch size, but there is no MultistepAdafactor.

>  is there an easy way to quantize the model for inference, for example float -> 8 bit quantization?

This was done in [Marian](https://github.com/marian-nmt/marian-dev). I am not sure if anything similar is possible in T2T. That said, you can use a T2T model as the teacher and a Marian model as the student.
